File: Sources/particles/Physics/SpatialGrid.swift
import Foundation
import Raylib

class SpatialGrid {
	private struct GridCell {
		var indices: [Int] = []
		var generation: Int = 0
	}

	private var grid: [GridCell]
	private var width: Int
	private var height: Int
	private var currentGeneration: Int = 0

	private func getGridIndex(_ x: Int, _ y: Int) -> Int {
		let clampedX = max(0, min(x, width - 1))
		let clampedY = max(0, min(y, height - 1))
		return clampedY * width + clampedX
	}

	init(width: Int, height: Int) {
		self.width = width
		self.height = height
		self.grid = Array(repeating: GridCell(), count: width * height)
	}

	func clear() {
		currentGeneration &+= 1
	}

	func insert(particleIdx: Int, pos: Vector2) {
		let cellX = Int(pos.x / CELL_SIZE)
		let cellY = Int(pos.y / CELL_SIZE)
		let idx = getGridIndex(cellX, cellY)

		if grid[idx].generation != currentGeneration {
			grid[idx].generation = currentGeneration
			grid[idx].indices.removeAll(keepingCapacity: true)
		}

		if grid[idx].indices.count < MAX_PARTICLES_PER_CELL {
			grid[idx].indices.append(particleIdx)
		}
	}

	func forNearby(pos: Vector2, callback: (Int) -> Void) {
		let cellX = Int(pos.x / CELL_SIZE)
		let cellY = Int(pos.y / CELL_SIZE)

		for dx in -1...1 {
			for dy in -1...1 {
				let idx = getGridIndex(cellX + dx, cellY + dy)
				if grid[idx].generation == currentGeneration {
					for particleIdx in grid[idx].indices {
						callback(particleIdx)
					}
				}
			}
		}
	}
}

File: src/spatial-grid.ts
import { CONFIG, GRID_HEIGHT, GRID_WIDTH } from "@/config.ts";

interface GridCell {
  indices: number[];
  generation: number;
}

export class SpatialGrid {
  private grid: GridCell[];
  private currentGeneration: number = 0;

  constructor() {
    this.grid = Array.from({ length: GRID_WIDTH * GRID_HEIGHT }, () => ({
      indices: [],
      generation: -1,
    }));
  }

  private getGridIndex(x: number, y: number): number {
    const clampedX = Math.max(0, Math.min(x, GRID_WIDTH - 1));
    const clampedY = Math.max(0, Math.min(y, GRID_HEIGHT - 1));
    return clampedY * GRID_WIDTH + clampedX;
  }

  clear(): void {
    this.currentGeneration += 1;
  }

  insert(particleIdx: number, x: number, y: number): void {
    const cellX = Math.floor(x / CONFIG.CELL_SIZE);
    const cellY = Math.floor(y / CONFIG.CELL_SIZE);
    const idx = this.getGridIndex(cellX, cellY);

    if (this.grid[idx].generation !== this.currentGeneration) {
      this.grid[idx].generation = this.currentGeneration;
      this.grid[idx].indices.length = 0;
    }

    if (this.grid[idx].indices.length < CONFIG.MAX_PARTICLES_PER_CELL) {
      this.grid[idx].indices.push(particleIdx);
    }
  }

  forNearby(x: number, y: number, callback: (idx: number) => void): void {
    const cellX = Math.floor(x / CONFIG.CELL_SIZE);
    const cellY = Math.floor(y / CONFIG.CELL_SIZE);

    for (let dx = -1; dx <= 1; dx++) {
      for (let dy = -1; dy <= 1; dy++) {
        const idx = this.getGridIndex(cellX + dx, cellY + dy);
        if (this.grid[idx].generation === this.currentGeneration) {
          for (const particleIdx of this.grid[idx].indices) {
            callback(particleIdx);
          }
        }
      }
    }
  }
}
